cloud/scope/powervs_cluster.go

```diff
@@ -1129,6 +1169,390 @@ func (s *PowerVSClusterScope) createVPCSubnet(subnet infrav1beta2.Subnet) (*stri
 	return subnetDetails.ID, nil
 }
 
+// ReconcileVPCSecurityGroup reconciles VPC security group.
+func (s *PowerVSClusterScope) ReconcileVPCSecurityGroup() (bool, error) {
+	for _, securityGroup := range s.IBMPowerVSCluster.Spec.VPCSecurityGroups {
+		var securityGroupID *string
+		var securityGroupRuleIDs []*string
+
+		if securityGroup.Name != nil {
+			securityGroupID, securityGroupRuleIDs = s.GetVPCSecurityGroupID(*securityGroup.Name)
+		} else {
+			securityGroupRuleIDs = s.GetVPCSecurityGroupRuleIDs(*securityGroup.ID)
+		}
+
+		if securityGroupID != nil && securityGroupRuleIDs != nil {
+			if _, _, err := s.IBMVPCClient.GetSecurityGroup(&vpcv1.GetSecurityGroupOptions{
+				ID: securityGroupID,
+			}); err != nil {
+				return true, err
+			}
+			for _, rule := range securityGroupRuleIDs {
+				if _, _, err := s.IBMVPCClient.GetSecurityGroupRule(&vpcv1.GetSecurityGroupRuleOptions{
+					SecurityGroupID: securityGroupID,
+					ID:              rule,
+				}); err != nil {
+					return true, err
```

Comment:
@dharaneeshvrd, shouldn't we be returning `false` on err and not requeue?

Reply:
The requeue was added for handling the [state of the infra resources](https://github.com/kubernetes-sigs/cluster-api-provider-ibmcloud/pull/1693), it can be skipped if security group doesn't need state management.

Reply:
Yes Agree, corrected.
Thanks @Amulyam24 !

Reply:
@dharaneesh, since we are returning false every time, maybe we can just return err and remove the requeue check in the controller?

Reply:
Corrected!